hw/chisel_acc/src/main/scala/snax_acc/reshuffle/Reshuffle.scala

```diff
@@ -0,0 +1,255 @@
+class ReshufflerDataIO(params: ReshufflerParams) extends Bundle {
+
+  // a multi-data input, decoupled interface for handshake
+  val input_i =
+    Flipped(Decoupled(UInt((params.poolLaneLen * params.inputWidth).W)))
+
+  // a multi-data output, decoupled interface for handshake
+  val output_o = Decoupled(
+    UInt((params.poolLaneLen * params.outputWidth).W)
+  )
+
+}
+
+// Reshuffler input and output declaration
+class ReshufflerIO(params: ReshufflerParams) extends Bundle {
+  // the input data across different MAXPoolPEs shares the same control signal
+  val ctrl = Flipped(
+    Decoupled(Vec(params.readWriteCsrNum, UInt(params.sizeConfigWidth.W)))
+  )
+  // decoupled data ports
+  val data = new ReshufflerDataIO(params)
+
+  val performance_counter = Output(UInt(32.W))
+  val busy_o = Output(Bool())
+}
+
+// Reshuffler module
+class Reshuffler(params: ReshufflerParams)
+    extends Module
+    with RequireAsyncReset {
+  val io = IO(new ReshufflerIO(params))
+
+  // control csr registers for storing the control data
+  val ctrl_csr = Reg(new ReshufflerCtrl(params))
+
+  // transpose module
+  val transposeMux = Module(new TransposeMux(params))
+
+  // generating parallel MAXPoolPEs
+  val lane = Seq.fill(params.poolLaneLen)(Module(new MAXPoolPE(params)))
+
+  // result from different MAXPoolPEs
+  val result = Wire(
+    Vec(params.poolLaneLen, SInt(params.outputWidth.W))
+  )
+
+  val pe_input_fire = WireInit(0.B)
+  val pe_input_fire_counter = RegInit(0.U(params.sizeConfigWidth.W))
+
+  val pe_out_fire = WireInit(0.B)
+  val pe_out_valid = WireInit(0.B)
+  val pe_out_fire_counter = RegInit(0.U(params.sizeConfigWidth.W))
+
+  // reshuffle result data output
+  val reshuffle_output_fire = WireInit(0.B)
+  val write_counter = RegInit(0.U(params.sizeConfigWidth.W))
+
+  // State declaration
+  val sIDLE :: sBUSY :: Nil = Enum(2)
+  val cstate = RegInit(sIDLE)
+  val nstate = WireInit(sIDLE)
+
+  // signals for state transition
+  val config_valid = WireInit(0.B)
+  val computation_finish = WireInit(0.B)
+
+  // Changing states
+  cstate := nstate
+
+  chisel3.dontTouch(cstate)
+  switch(cstate) {
+    is(sIDLE) {
+      when(config_valid) {
+        nstate := sBUSY
+      }.otherwise {
+        nstate := sIDLE
+      }
+
+    }
+    is(sBUSY) {
+      when(computation_finish) {
+        nstate := sIDLE
+      }.otherwise {
+        nstate := sBUSY
+      }
+    }
+  }
+
+  io.busy_o := cstate === sBUSY
+
+  val performance_counter = RegInit(0.U(32.W))
+  when(cstate === sBUSY) {
+    performance_counter := performance_counter + 1.U
+  }.elsewhen(config_valid) {
+    performance_counter := 0.U
+  }
+  io.performance_counter := performance_counter
+
+  config_valid := io.ctrl.fire
+
+  // when config valid, store the configuration for later computation
+  when(config_valid) {
+    ctrl_csr.opcode_i := io.ctrl.bits(0)(1, 0)
+    ctrl_csr.reduceLen_i := io.ctrl.bits(0)(6, 2)
+    ctrl_csr.tloop2Len_i := io.ctrl.bits(0)(31, 7)
```

Comment:
Wait there are more features??? haha what are reduce len? what is also tloop? XD

Maybe give minimal documentation to PR on what is expected of this accelerator?